Explain tempdata (#14127)
aspnet/AspNetCore.Docs: aspnetcore/fundamentals/app-state.md

## TempData

ASP.NET Core exposes the Razor Pages [TempData](xref:Microsoft.AspNetCore.Mvc.RazorPages.PageModel.TempData) or Controller <xref:Microsoft.AspNetCore.Mvc.Controller.TempData>. This property stores data until it's read in another request. [Keep(String)](xref:Microsoft.AspNetCore.Mvc.ViewFeatures.ITempDataDictionary.Keep*) and [Peek(string)](xref:Microsoft.AspNetCore.Mvc.ViewFeatures.ITempDataDictionary.Peek*) methods can be used to examine the data without deletion at the end of the request. [Keep()](xref:Microsoft.AspNetCore.Mvc.ViewFeatures.ITempDataDictionary.Keep*) marks all items in the dictionary for retention. `TempData` is particularly useful for redirection when data is required for more than a single request. `TempData` is implemented by `TempData` providers using either cookies or session state.

## TempData samples

Consider the following page that creates a customer:

> [!code-csharp[](app-state/3.0samples/RazorPagesContacts/Pages/Customers/Create.cshtml.cs?name=snippet&highlight=15-16,30)]

Reviewer: The naming convention for this samples folder deviates from the convention used elsewhere. I suggest creating a 3.x folder within app-state/samples/.

The following page displays `TempData["Message"]`:

> [!code-cshtml[](app-state/3.0samples/RazorPagesContacts/Pages/Customers/IndexPeek.cshtml?range=1-14)]

Reviewer: Consider using a named region in the CSHTML file instead of a brittle range. Here's an example of the syntax in the CSHTML file:

```
<!-- <snippet_IndexPeek> -->
markup here
<!-- </snippet_IndexPeek> -->
```

Author: That doesn't work
![image](https://user-images.githubusercontent.com/3605364/64371787-4c34f200-cfd6-11e9-85af-4b09fb904631.png)
And the following is not recognized as a snippet

```
@*<snippet>*@
@page
@model IndexModel

<h1>Contacts Keep</h1>

@{
    if (TempData["Message"] != null)
    {
        <h3>Message: @TempData["Message"]</h3>
    }
    TempData.Keep("Message");
}

@*Content removed for brevity.*@

@*</snippet>*@
```
Your approach works except when you need the first line `@page`

Author: BTW, I opened a bug asking for `@*<snippet>*@` to work - over a year ago.

Reviewer: Looks like this is one edge case where the snippet syntax won't work. That's only because of the requirement that the `@page` directive must appear at the top.

Author: I wouldn't call it an edge case as you frequently need that line. I'll check up on the bug.

In the preceding markup, at the end of the request, `TempData["Message"]` is **not** deleted because `Peek` is used. Refreshing the page displays `TempData["Message"]`.

> The following markup is similar to the preceding code, but uses `Keep` to mark the data not to be deleted at the end of the request:

Reviewer: ```suggestion
The following markup is similar to the preceding code, but uses `Keep` to preserve the data at the end of the request:
```

[!code-cshtml[](app-state/3.0samples/RazorPagesContacts/Pages/Customers/IndexKeep.cshtml?range=1-14)]

> Navigating between the IndexPeek an IndexKeep pages will not delete `TempData["Message"]`.

Reviewer: ```suggestion
Navigating between the *IndexPeek* and *IndexKeep* pages won't delete `TempData["Message"]`.
```

The following code displays `TempData["Message"]`, but at the end of the request, `TempData["Message"]` is deleted:

> [!code-cshtml[](app-state/3.0samples/RazorPagesContacts/Pages/Customers/Index.cshtml?range=1-14)]

Reviewer: Consider using a named region here too.
